### ISSUE TITLE
Phantom sometimes cannot connect to Cassandra with ConnectionException

### ISSUE THREAD

**Issue by GRpro** [outworkers/phantom#744]
Sometimes phantom cannot connect to Cassandra host on start. I have a single Cassandra node running locally with default cassandra.yml properties. 
```
$ cassandra -v
3.11.0
```

phantom version:
```
libraryDependencies += "com.outworkers" %% "phantom-dsl" % "2.12.1"
```

Connector creation:
```
  lazy val Connector = ContactPoints(Seq(Configuration.CassandraHost), Configuration.CassandraPort)
    .withClusterBuilder(
      _.withSocketOptions(
        new SocketOptions()
          .setReadTimeoutMillis(10000)
          .setConnectTimeoutMillis(20000))
        .withPoolingOptions(
          new PoolingOptions()
            .setMaxQueueSize(100000)
            .setPoolTimeoutMillis(20000))
        .withRetryPolicy(new MultipleRetryPolicy(5))
        .withReconnectionPolicy(new ConstantReconnectionPolicy(3000L))
    ).keySpace(CassandraKeyspace)

```
Find logs below.

```
2017-08-27 15:41:14,574 [DEBUG] c.d.d.c.H.STATES - [localhost/0:0:0:0:0:0:0:1:9042] Connection[localhost/0:0:0:0:0:0:0:1:9042-1, inFlight=0, closed=false] Transport initialized, connection ready
2017-08-27 15:41:14,575 [DEBUG] c.d.d.c.ControlConnection - [Control connection] Refreshing node list and token map
2017-08-27 15:41:14,643 [WARN] c.d.d.c.Cluster - You listed localhost/127.0.0.1:9042 in your contact points, but it wasn't found in the control host's system.peers at startup
2017-08-27 15:41:14,649 [DEBUG] c.d.d.c.ControlConnection - [Control connection] Refreshing schema
2017-08-27 15:41:14,779 [DEBUG] c.d.d.c.H.STATES - [Control connection] established to localhost/0:0:0:0:0:0:0:1:9042
2017-08-27 15:41:14,779 [INFO] c.d.d.c.p.DCAwareRoundRobinPolicy - Using data-center name 'datacenter1' for DCAwareRoundRobinPolicy (if this is incorrect, please provide the correct datacenter name with DCAwareRoundRobinPolicy constructor)
2017-08-27 15:41:14,781 [INFO] c.d.d.c.Cluster - New Cassandra host localhost/0:0:0:0:0:0:0:1:9042 added
2017-08-27 15:41:14,781 [INFO] c.d.d.c.Cluster - New Cassandra host /172.25.0.2:9042 added
2017-08-27 15:41:14,784 [DEBUG] c.d.d.c.SystemProperties - com.datastax.driver.CHECK_IO_DEADLOCKS is undefined, using default value true
2017-08-27 15:41:14,786 [DEBUG] c.d.d.c.H.STATES - [localhost/0:0:0:0:0:0:0:1:9042] preparing to open 1 new connections, total = 2
2017-08-27 15:41:14,787 [DEBUG] c.d.d.c.Connection - Connection[localhost/0:0:0:0:0:0:0:1:9042-2, inFlight=0, closed=false] Connection established, initializing transport
2017-08-27 15:41:14,791 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] preparing to open 1 new connections, total = 1
2017-08-27 15:41:14,794 [DEBUG] c.d.d.c.H.STATES - [localhost/0:0:0:0:0:0:0:1:9042] Connection[localhost/0:0:0:0:0:0:0:1:9042-2, inFlight=0, closed=false] Transport initialized, connection ready
2017-08-27 15:41:14,795 [DEBUG] c.d.d.c.HostConnectionPool - Created connection pool to host localhost/0:0:0:0:0:0:0:1:9042 (1 connections needed, 1 successfully opened)
2017-08-27 15:41:14,795 [DEBUG] c.d.d.c.Session - Added connection pool for localhost/0:0:0:0:0:0:0:1:9042
2017-08-27 15:41:34,796 [DEBUG] c.d.d.c.Connection - Connection[/172.25.0.2:9042-1, inFlight=0, closed=false] Error connecting to /172.25.0.2:9042 (connection timed out: /172.25.0.2:9042)
2017-08-27 15:41:34,798 [DEBUG] c.d.d.c.H.STATES - Defuncting Connection[/172.25.0.2:9042-1, inFlight=0, closed=false] because: [/172.25.0.2:9042] Cannot connect
2017-08-27 15:41:34,798 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] preventing new connections for the next 3000 ms
2017-08-27 15:41:34,798 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] Connection[/172.25.0.2:9042-1, inFlight=0, closed=false] failed, remaining = 0
2017-08-27 15:41:34,800 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] marking host DOWN
2017-08-27 15:41:34,800 [DEBUG] c.d.d.c.Connection - Connection[/172.25.0.2:9042-1, inFlight=0, closed=true] closing connection
2017-08-27 15:41:34,802 [DEBUG] c.d.d.c.AbstractReconnectionHandler - First reconnection scheduled in 3000ms
2017-08-27 15:41:34,803 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] starting reconnection attempt
2017-08-27 15:41:34,806 [WARN] c.d.d.c.Session - Error creating pool to /172.25.0.2:9042
com.datastax.driver.core.exceptions.ConnectionException: [/172.25.0.2:9042] Pool was closed during initialization
	at com.datastax.driver.core.HostConnectionPool$2.onSuccess(HostConnectionPool.java:148)
	at com.datastax.driver.core.HostConnectionPool$2.onSuccess(HostConnectionPool.java:134)
	at com.google.common.util.concurrent.Futures$6.run(Futures.java:1773)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:456)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:817)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:753)
	at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:613)
	at com.google.common.util.concurrent.CollectionFuture$CollectionFutureRunningState.handleAllCompleted(CollectionFuture.java:76)
	at com.google.common.util.concurrent.AggregateFuture$RunningState.processCompleted(AggregateFuture.java:255)
	at com.google.common.util.concurrent.AggregateFuture$RunningState.decrementCountAndMaybeComplete(AggregateFuture.java:242)
2017-08-27 15:41:34,810 [INFO] c.o.p.c.DefaultSessionProvider - Automatically initialising keyspace recommender_cf with query CREATE KEYSPACE IF NOT EXISTS recommender_cf WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 1};
2017-08-27 15:41:35,845 [DEBUG] c.o.phantom - Executing query: CREATE TABLE IF NOT EXISTS recommender_cf.preference (key text, userId text, itemId text, rating double, timestamp bigint, PRIMARY KEY (key));
2017-08-27 15:41:35,855 [DEBUG] c.o.phantom - Creating secondary indexes on recommender_cf.preference
2017-08-27 15:41:35,863 [INFO] g.m.a.b.ApplicationLifeCycle - recommender database created
2017-08-27 15:41:36,115 [INFO] a.e.s.Slf4jLogger - Slf4jLogger started
2017-08-27 15:41:36,263 [DEBUG] c.t.s.a.AkkaSSLConfig - Initializing AkkaSSLConfig extension...
2017-08-27 15:41:36,266 [DEBUG] c.t.s.a.AkkaSSLConfig - buildHostnameVerifier: created hostname verifier: com.typesafe.sslconfig.ssl.DefaultHostnameVerifier@5762658b
2017-08-27 15:41:37,003 [INFO] recommender-service - Bound to /0:0:0:0:0:0:0:0:7654 
2017-08-27 15:41:37,013 [DEBUG] default/Pool(shared->http://localhost:9000) - (Re-)starting host connection pool to localhost:9000
2017-08-27 15:41:37,135 [DEBUG] default/Pool(shared->http://localhost:9000) - InputBuffer (max-open-requests = 32) now filled with 1 request after enqueuing POST /api/recommenders Strict(931 bytes)
2017-08-27 15:41:37,143 [DEBUG] default/Pool(shared->http://localhost:9000) - [0] Unconnected -> Busy(1)
2017-08-27 15:41:37,172 [DEBUG] default/Pool(shared->http://localhost:9000) - [0] <unconnected> Establishing connection...
2017-08-27 15:41:37,209 [DEBUG] default/Pool(shared->http://localhost:9000) - [0] <unconnected> pushing request to connection: POST /api/recommenders Strict(931 bytes)
2017-08-27 15:41:37,302 [DEBUG] default/Pool(shared->http://localhost:9000) - [0] </127.0.0.1:50480->localhost:9000> Received response: POST /api/recommenders Strict(931 bytes) -> 200 OK Empty
2017-08-27 15:41:37,306 [DEBUG] default/Pool(shared->http://localhost:9000) - [0] </127.0.0.1:50480->localhost:9000> Finished reading response entity for POST /api/recommenders Strict(931 bytes) -> 200 OK Empty
2017-08-27 15:41:37,307 [INFO] g.m.a.b.ApplicationLifeCycle - Registered recommender in web_console
2017-08-27 15:41:37,310 [DEBUG] default/Pool(shared->http://localhost:9000) - [0] Busy(1) -> Idle
2017-08-27 15:41:37,805 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] preparing to open 1 new connections, total = 1
2017-08-27 15:41:44,683 [DEBUG] c.d.d.c.Connection - Connection[localhost/0:0:0:0:0:0:0:1:9042-1, inFlight=0, closed=false] was inactive for 30 seconds, sending heartbeat
2017-08-27 15:41:44,686 [DEBUG] c.d.d.c.Connection - Connection[localhost/0:0:0:0:0:0:0:1:9042-1, inFlight=0, closed=false] heartbeat query succeeded
2017-08-27 15:41:57,810 [DEBUG] c.d.d.c.Connection - Connection[/172.25.0.2:9042-2, inFlight=0, closed=false] Error connecting to /172.25.0.2:9042 (connection timed out: /172.25.0.2:9042)
2017-08-27 15:41:57,810 [DEBUG] c.d.d.c.H.STATES - Defuncting Connection[/172.25.0.2:9042-2, inFlight=0, closed=false] because: [/172.25.0.2:9042] Cannot connect
2017-08-27 15:41:57,810 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] Connection[/172.25.0.2:9042-2, inFlight=0, closed=false] failed, remaining = 0
2017-08-27 15:41:57,811 [DEBUG] c.d.d.c.Connection - Connection[/172.25.0.2:9042-2, inFlight=0, closed=true] closing connection
2017-08-27 15:41:57,811 [DEBUG] c.d.d.c.Cluster - Aborting onDown because a reconnection is running on DOWN host /172.25.0.2:9042
2017-08-27 15:41:57,812 [DEBUG] c.d.d.c.Cluster - Failed reconnection to /172.25.0.2:9042 ([/172.25.0.2:9042] Cannot connect), scheduling retry in 3000 milliseconds
2017-08-27 15:41:57,812 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] next reconnection attempt in 3000 ms
2017-08-27 15:42:00,813 [DEBUG] c.d.d.c.H.STATES - [/172.25.0.2:9042] preparing to open 1 new connections, total = 1
2017-08-27 15:42:05,855 [DEBUG] c.d.d.c.Connection - Connection[localhost/0:0:0:0:0:0:0:1:9042-2, inFlight=0, closed=false] was inactive for 30 seconds, sending heartbeat
2017-08-27 15:42:05,858 [DEBUG] c.d.d.c.Connection - Connection[localhost/0:0:0:0:0:0:0:1:9042-2, inFlight=0, closed=false] heartbeat query succeeded
```

**Comment by hsn10** [comment]
I seen this message _Pool was closed during initialization_ too, but it did not had any negative effect. Application still works fine, no exception is seen from client code. So I assume, its not related  to later connection problem.

Your problem is in Cassandra Java driver itself, not Phantom. It needs to be reported to DataStax bug tracker.

**Comment by alexflav23** [comment]
Hi,

Might want to check networking and timeouts first, looks like a connection timeout explicitly happening for whatever reason:

`c.d.d.c.Connection - Connection[/172.25.0.2:9042-2, inFlight=0, closed=false] Error connecting to /172.25.0.2:9042 (connection timed out: /172.25.0.2:9042)`.

**Comment by GRpro** [comment]
Thank you!
